src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs

```diff
@@ -85,25 +81,26 @@ public ISingleFileComponentRecorder CreateSingleFileComponentRecorder(string loc
             throw new ArgumentNullException(nameof(location));
         }
 
-        var matching = this.singleFileRecorders.FirstOrDefault(x => x.ManifestFileLocation == location);
-        if (matching == null)
-        {
-            matching = new SingleFileComponentRecorder(location, this, this.enableManualTrackingOfExplicitReferences, this.logger);
-            this.singleFileRecorders.Add(matching);
-        }
-
-        return matching;
+        return this.singleFileRecorders.GetOrAdd(location, loc => new SingleFileComponentRecorder(loc, this, this.enableManualTrackingOfExplicitReferences, this.logger));
     }
 
     public IReadOnlyDictionary<string, IDependencyGraph> GetDependencyGraphsByLocation()
     {
-        return this.singleFileRecorders.Where(x => x.DependencyGraph.HasComponents())
-            .ToImmutableDictionary(x => x.ManifestFileLocation, x => x.DependencyGraph as IDependencyGraph);
+        return new ReadOnlyDictionary<string, IDependencyGraph>(
+            this.singleFileRecorders.Values
+                .Where(x => x.DependencyGraph.HasComponents())
+                .ToDictionary(x => x.ManifestFileLocation, x => x.DependencyGraph as IDependencyGraph));
     }
 
     internal DependencyGraph GetDependencyGraphForLocation(string location)
     {
-        return this.singleFileRecorders.Single(x => x.ManifestFileLocation == location).DependencyGraph;
+        if (this.singleFileRecorders.TryGetValue(location, out var singleFileRecorder))
+        {
+            return singleFileRecorder.DependencyGraph;
+        }
+
+        // TODO: Should we throw an exception here instead of returning null?
```

Comment:
I'd say yes. Previous behavior of .Single() was to throw an InvalidOperationException if nothing matched